File: app-android/app/src/main/java/com/aengix/hopper/ssh/DeploySSH.kt
package com.aengix.hopper.ssh

import com.aengix.hopper.model.HopConstants
import com.aengix.hopper.util.IPv4Only
import com.aengix.hopper.util.ShellQuote
import net.schmizz.sshj.SSHClient
import net.schmizz.sshj.transport.verification.PromiscuousVerifier

object DeploySSH {
    fun runInstall(
        host: String,
        port: Int,
        user: String,
        password: String?,
        privateKeyPem: String,
        publicKeyLine: String,
        installDir: String,
        onLog: ((String) -> Unit)? = null,
    ): String {
        val log = onLog ?: {}
        log("Connecting to $user@$host:$port…")

        HopSecurityProviders.ensureRegistered()
        val resolvedHost = IPv4Only.resolveHost(host)
        val sshClient = SSHClient()
        sshClient.addHostKeyVerifier(PromiscuousVerifier())
        sshClient.timeout = 120_000
        try {
            sshClient.connect(resolvedHost, port)
            if (password != null) {
                sshClient.authPassword(user, password)
            } else {
                val keyProvider = HopKeyProviders.keyProviderFor(privateKeyPem)
                sshClient.authPublickey(user, keyProvider)
            }

            log("Ensuring deploy key is in authorized_keys…")
            HopSSH.runCommand(sshClient, authorizedKeysCommand(publicKeyLine, host))
            log("Deploy key authorized.")

            log("Running remote install…")
            return HopSSH.runCommand(sshClient, installCommand(host, port, installDir), onLine = log)
        } finally {
            runCatching { sshClient.disconnect() }
        }
    }

    private fun authorizedKeysCommand(publicKeyLine: String, host: String): String {
        val pub = ShellQuote.bashSingle(publicKeyLine)
        return "mkdir -p ~/.ssh && chmod 700 ~/.ssh && touch ~/.ssh/authorized_keys && chmod 600 ~/.ssh/authorized_keys && " +
            "grep -qF $pub ~/.ssh/authorized_keys 2>/dev/null || echo $pub deploy-${ShellQuote.bashSingle(host)} >>~/.ssh/authorized_keys"
    }

    private fun installCommand(host: String, port: Int, installDir: String): String {
        val url = ShellQuote.bashSingle(HopConstants.SERVER_INSTALL_URL)
        val ref = ShellQuote.bashSingle(HopConstants.SERVER_INSTALL_REF)
        val dir = ShellQuote.bashSingle(installDir)
        val hostQ = ShellQuote.bashSingle(host)
        val portQ = ShellQuote.bashSingle(port.toString())
        return "curl -fsSL $url | HOPPER_REF=$ref HOPPER_INSTALL_DIR=$dir bash -s -- --configure --host $hostQ --port $portQ"
    }
}

File: app/TunnelCore/DeploySSH.swift
import Citadel
import Crypto
import Foundation
import NIOSSH

enum DeploySSH {
    static func runInstall(
        host: String,
        port: Int,
        user: String,
        password: String?,
        privateKeyPEM: String,
        publicKeyLine: String,
        installDir: String
    ) async throws -> String {
        let client: SSHClient
        if let password {
            client = try await connect(host: host, port: port, user: user, password: password)
        } else {
            client = try await connect(host: host, port: port, user: user, privateKeyPEM: privateKeyPEM)
        }

        defer { Task { try? await client.close() } }

        let authCmd = authorizedKeysCommand(publicKeyLine: publicKeyLine, host: host)
        _ = try await HopSSH.runCommand(on: client, authCmd)

        let installCmd = installCommand(
            host: host,
            port: port,
            installDir: installDir
        )
        return try await HopSSH.runCommand(on: client, installCmd)
    }

    private static func connect(host: String, port: Int, user: String, password: String) async throws -> SSHClient {
        var settings = SSHClientSettings(
            host: host,
            port: port,
            authenticationMethod: { .passwordBased(username: user, password: password) },
            hostKeyValidator: .acceptAnything()
        )
        settings.group = HopSSH.sharedEventLoop
        settings.connectTimeout = .seconds(60)
        settings.loginTimeout = .seconds(60)
        return try await SSHClient.connect(to: settings)
    }

    private static func connect(host: String, port: Int, user: String, privateKeyPEM: String) async throws -> SSHClient {
        let privateKey = try Curve25519.Signing.PrivateKey(sshEd25519: privateKeyPEM)
        var settings = SSHClientSettings(
            host: host,
            port: port,
            authenticationMethod: { .ed25519(username: user, privateKey: privateKey) },
            hostKeyValidator: .acceptAnything()
        )
        settings.group = HopSSH.sharedEventLoop
        settings.connectTimeout = .seconds(60)
        settings.loginTimeout = .seconds(60)
        return try await SSHClient.connect(to: settings)
    }

    private static func authorizedKeysCommand(publicKeyLine: String, host: String) -> String {
        let pub = ShellQuote.bashSingle(publicKeyLine)
        return """
        mkdir -p ~/.ssh && chmod 700 ~/.ssh && touch ~/.ssh/authorized_keys && chmod 600 ~/.ssh/authorized_keys && grep -qF \(pub) ~/.ssh/authorized_keys 2>/dev/null || echo \(pub) deploy-\(ShellQuote.bashSingle(host)) >>~/.ssh/authorized_keys
        """
    }

    private static func installCommand(host: String, port: Int, installDir: String) -> String {
        let url = ShellQuote.bashSingle(HopConstants.serverInstallURL)
        let ref = ShellQuote.bashSingle(HopConstants.serverInstallRef)
        let dir = ShellQuote.bashSingle(installDir)
        let hostQ = ShellQuote.bashSingle(host)
        let portQ = ShellQuote.bashSingle(String(port))
        return "curl -fsSL \(url) | HOPPER_REF=\(ref) HOPPER_INSTALL_DIR=\(dir) bash -s -- --configure --host \(hostQ) --port \(portQ)"
    }
}
